[16.0] [MIG] account_fiscal_year_closing

Borruso opened the pull request:
Migration account_fiscal_year_closing from 14.0 to 16.0

--
Confermo di aver firmato il CLA https://odoo-community.org/page/cla e di aver letto le linee guida su https://odoo-community.org/page/contributing

antespi committed:
[ADD] account_fiscalyear_closing: Generic engine for closing operations

pedrobaeza committed:
[FIX+IMP] account_fiscal_year_closing

* Included template mechanism
* Remove unused files
* Fix behaviors
* Some refactoring
* Complete features

JordiBForgeFlow committed:
migrate account_fiscal_year_closing to 10.0

sergiocorato committed:
[12.0][PORT] account fiscal year closing

alexey-pelykh committed:
[FIX] account_fiscal_year_closing: tests should not use fixed dates

eLBati committed:
Translated using Weblate (Italian)

Currently translated at 66.6% (76 of 114 strings)

Translation: account-closing-12.0/account-closing-12.0-account_fiscal_year_closing
Translate-URL: https://translation.odoo-community.org/projects/account-closing-12-0/account-closing-12-0-account_fiscal_year_closing/it/

eLBati committed:
IMP account_fiscal_year_closing adding multi company rule for account.fiscalyear.closing.template

marcelsavegnago committed:
Added translation using Weblate (Portuguese (Brazil))

SimoRubi committed:
[MIG] account_fiscal_year_closing: Migration to 14.0
Fixed warnings:
Field account.fiscalyear.closing.closing_template_id: unknown parameter 'oldname', if this is an actual parameter you may want to override the method _valid_field_parameter on the relevant model in order to allow it
Field account.move.fyc_id: unknown parameter 'delete', if this is an actual parameter you may want to override the method _valid_field_parameter on the relevant model in order to allow it
The model account.fiscalyear.closing.config has no _description
The model account.fiscalyear.closing.mapping has no _description
The model account.fiscalyear.closing.type has no _description
The model account.fiscalyear.closing.template has no _description
The model account.fiscalyear.closing.config.template has no _description
The model account.fiscalyear.closing.mapping.template has no _description
The model account.fiscalyear.closing.type.template has no _description
The model account.fiscalyear.closing.unbalanced.move has no access rules
The model account.fiscalyear.closing.unbalanced.move.line has no access rules
odoo.addons.base.models.ir_ui_view: Search tag requires at least one field element
The method '_company_default_get' on res.company is deprecated and shouldn't be used anymore
DeprecationWarning: RedirectWarning method 'post()' is a deprecated alias to 'action_post()' or _post()

SimoRubi committed:
[IMP] account_fiscal_year_closing: black, isort, prettier

SimoRubi committed:
[IMP] account_fiscal_year_closing: Restore onchange on closing's template

SimoRubi committed:
[IMP] account_fiscal_year_closing: Add date field to unbalanced move line.
Otherwise the following might happen:
  File "/path/to/odoo/addons/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 328, in button_calculate
    res = self.calculate()
  File "/path/to/odoo/addons/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 315, in calculate
    return self._show_unbalanced_move_wizard(data)
  File "/path/to/odoo/addons/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 295, in _show_unbalanced_move_wizard
    wizard = self.env["account.fiscalyear.closing.unbalanced.move"].create(data)
[...]
ValueError: Invalid field 'date' on model 'account.fiscalyear.closing.unbalanced.move.line'

Commit:
[FIX][account_fiscal_year_closing] fix create and write

alessandrocamilli committed:
Translated using Weblate (Italian)

Currently translated at 92.1% (105 of 114 strings)

Translation: account-closing-14.0/account-closing-14.0-account_fiscal_year_closing
Translate-URL: https://translation.odoo-community.org/projects/account-closing-14-0/account-closing-14-0-account_fiscal_year_closing/it/

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 92.1% (105 of 114 strings)

Translation: account-closing-14.0/account-closing-14.0-account_fiscal_year_closing
Translate-URL: https://translation.odoo-community.org/projects/account-closing-14-0/account-closing-14-0-account_fiscal_year_closing/it/

primes2h committed:
Translated using Weblate (Italian)

Currently translated at 92.9% (106 of 114 strings)

Translation: account-closing-14.0/account-closing-14.0-account_fiscal_year_closing
Translate-URL: https://translation.odoo-community.org/projects/account-closing-14-0/account-closing-14-0-account_fiscal_year_closing/it/

alessandrocamilli committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (114 of 114 strings)

Translation: account-closing-14.0/account-closing-14.0-account_fiscal_year_closing
Translate-URL: https://translation.odoo-community.org/projects/account-closing-14-0/account-closing-14-0-account_fiscal_year_closing/it/

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (114 of 114 strings)

Translation: account-closing-14.0/account-closing-14.0-account_fiscal_year_closing
Translate-URL: https://translation.odoo-community.org/projects/account-closing-14-0/account-closing-14-0-account_fiscal_year_closing/it/

michelerusti committed:
[14.0][FIX] - account_fiscal_year_closing - closing procedure computes canceled entries in closing

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (114 of 114 strings)

Translation: account-closing-14.0/account-closing-14.0-account_fiscal_year_closing
Translate-URL: https://translation.odoo-community.org/projects/account-closing-14-0/account-closing-14-0-account_fiscal_year_closing/it/

douglascstd committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 95.6% (109 of 114 strings)

Translation: account-closing-14.0/account-closing-14.0-account_fiscal_year_closing
Translate-URL: https://translation.odoo-community.org/projects/account-closing-14-0/account-closing-14-0-account_fiscal_year_closing/pt_BR/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (114 of 114 strings)

Translation: account-closing-14.0/account-closing-14.0-account_fiscal_year_closing
Translate-URL: https://translation.odoo-community.org/projects/account-closing-14-0/account-closing-14-0-account_fiscal_year_closing/it/

Borruso committed:
[IMP] account_fiscal_year_closing: pre-commit stuff

Borruso committed:
[MIG] account_fiscal_year_closing: Migration to 16.0

ghost commented:
Will this be merged?

SirAionTech commented on account_fiscal_year_closing/models/account_fiscalyear_closing_abstract.py:
```diff
@@ -81,8 +81,27 @@ class AccountFiscalyearClosingTypeAbstract(models.AbstractModel):
-    account_type_id = fields.Many2one(
-        comodel_name="account.account.type",
-        string="Account type",
+    account_type = fields.Selection(
```
This change requires a migration, otherwise when a database is migrated the existing records will have an empty `account_type` instead of having the value corresponding to their old `account_type_id`.
You can have a look at https://oca.github.io/OpenUpgrade for better understanding what a migration is.

Since the migration is usually not trivial and it does not really affect the module's behavior, sometimes this can be done later; this will be probably decided by whoever wants to merge this module.

If this is the case, please at least write a note in the roadmap of the module.

Borruso replied:
create migration
await you feedback

SirAionTech replied:
:+1:

SirAionTech commented on account_fiscal_year_closing/tests/test_account_fiscal_year_closing.py:
```diff
@@ -0,0 +1,394 @@
+                    not in [
+                        "asset_receivable",  # Receivable
+                        "asset_current",  # Current Assets
+                        "income",  # Current Assets
```
```suggestion
                        "income",  # Income
```

Borruso replied:
done

SirAionTech replied:
:+1:

SirAionTech reviewed (changes requested):
Thanks for the PR!
Please squash the commits as described in https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests#mergesquash-the-commits-generated-by-bots-or-weblate.

SirAionTech reviewed (changes requested):
Commit https://github.com/OCA/account-closing/commit/21851e97d7bd26b8c4716960ca9689ff395ff152 is missing from history

Borruso commented:
> Commit [21851e9](https://github.com/OCA/account-closing/commit/21851e97d7bd26b8c4716960ca9689ff395ff152) is missing from history

done

gjlong68 commented:
V16
![image](https://github.com/OCA/account-closing/assets/52815201/2f739109-f887-4fcc-8607-1fb210a6f8d4)

Borruso commented:
> V16 ![image](https://user-images.githubusercontent.com/52815201/276044172-2f739109-f887-4fcc-8607-1fb210a6f8d4.png)

@gjlong68, can you check again, please?

GSLabIt commented on account_fiscal_year_closing/security/account_fiscalyear_closing_security.xml:
```diff
@@ -0,0 +1,24 @@
+        <field name="global" eval="True" />
+        <field
+            name="domain_force"
+        >['|', ('company_id', '=', False), ('company_id', 'child_of', [user.company_id.id])]</field>
```
```suggestion
        >['|', ('company_id', '=', False), ('company_id', 'in', company_ids)]</field>
```

GSLabIt commented on account_fiscal_year_closing/security/account_fiscalyear_closing_security.xml:
```diff
@@ -0,0 +1,24 @@
+        <field name="global" eval="True" />
+        <field
+            name="domain_force"
+        >['|', ('company_id', '=', False), ('company_id', 'child_of', [user.company_id.id])]</field>
```
```suggestion
        >['|', ('company_id', '=', False), ('company_id', 'in', company_ids)]</field>
```

GSLabIt reviewed (commented):
multi company fix

Borruso commented:
> multi company fix

@GSLabIt  thanks for suggest
added migration

gjlong68 commented:
Borruso:OK

mide-dynapps commented:
Hi guys, thanks for your job.

Has anyone tested the migration script?


On my side, it's KO (I'm on Odoo.sh), the pre-mig script doesn't work.
IDs, use with env.ref("account.data_account_type*), don't exist on the migrated system (when the pre script runs)

SirAionTech commented:
> Has anyone tested the migration script?

I didn't

> On my side, it's KO (I'm on Odoo.sh), the pre-mig script doesn't work.
IDs, use with env.ref("account.data_account_type*), don't exist on the migrated system (when the pre script runs)

Then please add that as a review to the PR, have a look at https://odoo-community.org/resources/review if you don't know how.

pedrobaeza committed:
[REF] account_fiscal_year_closing: Debrand module

Although we originally developped the module, we figured out soon
that this approach is not correct, so we prefer to not appearing
in the credits.

pedrobaeza commented:
Please include 55998e8fc84b93764a66df09347467ae19add7e3 in this PR.

pedrobaeza commented:
/ocabot migration account_fiscal_year_closing

lk-eska commented:
These 2 PR's are will be necessary as well:

https://github.com/OCA/account-closing/pull/288
https://github.com/OCA/account-closing/pull/294

Borruso committed:
[FIX] account_fiscal_year_closing: move line created with correct currency

Borruso committed:
[FIX] account_fiscal_year_closing: create/write
When creating from scratch, Mapping is used as an Array instead of simple id (due to many in another many).
When coming from a template, id is used
To take into account this particularity, check the type before taking the id of the array.

lk-eska commented:
Thanks for the updates @Borruso

gjlong68 commented:
account_fiscal_year_closing(V17)?




郭Sir
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "OCA/account-closing"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年5月10日(星期五) 下午4:11
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [OCA/account-closing] [16.0] [MIG] account_fiscal_year_closing (PR #248)





 
These 2 PR's are will be necessary as well:
 
#288
 #294
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***&gt;

odooNextev commented:
Does it need to be approved from a PSC member before asking for merge?

OKT96 commented on account_fiscal_year_closing/migrations/16.0.1.0.0/pre-migration.py:
```diff
@@ -0,0 +1,64 @@
+def migrate(env, version):
+
+    all_account_type = [
+        ("asset_receivable", env.ref("account.data_account_type_receivable").id),
```
When migrating from v15 to v16 using OpenUpgrade, the following error occurs:
```
2024-07-30 09:16:17,113 2133688 INFO cbx_2024-07-30_v16 odoo.modules.loading: Loading module account_fiscal_year_closing (102/170)
2024-07-30 09:16:17,430 2133688 INFO cbx_2024-07-30_v16 odoo.modules.migration: module account_fiscal_year_closing: Running migration [>16.0.1.0.0] pre-migration
2024-07-30 09:16:17,444 2133688 INFO cbx_2024-07-30_v16 OpenUpgrade: account_fiscal_year_closing: pre-migration script called with version 15.0.1.0.0
2024-07-30 09:16:17,444 2133688 ERROR cbx_2024-07-30_v16 OpenUpgrade: account_fiscal_year_closing: error in migration script /opt/odoo/local/addons-oca-community/addons/account_fiscal_year_closing/migrations/16.0.1.0.0/pre-migration.py: 'account.account.type'
2024-07-30 09:16:17,444 2133688 ERROR cbx_2024-07-30_v16 OpenUpgrade: 'account.account.type'
Traceback (most recent call last):
  File "/srv/odoo/venv/lib/python3.10/site-packages/openupgradelib/openupgrade.py", line 2292, in wrapped_function
    func(
  File "/opt/odoo/local/addons-oca-community/addons/account_fiscal_year_closing/migrations/16.0.1.0.0/pre-migration.py", line 31, in migrate
    ("asset_receivable", env.ref("account.data_account_type_receivable").id),
  File "/opt/odoo/src/odoo/odoo/api.py", line 600, in ref
    record = self[res_model].browse(res_id)
  File "/opt/odoo/local/OpenUpgrade/openupgrade_framework/odoo_patch/odoo/api.py", line 39, in __getitem__
    return Environment.__getitem__._original_method(self, model_name)
  File "/opt/odoo/src/odoo/odoo/api.py", line 550, in __getitem__
    return self.registry[model_name](self, (), ())
  File "/opt/odoo/src/odoo/odoo/modules/registry.py", line 190, in __getitem__
    return self.models[model_name]
KeyError: 'account.account.type'
2024-07-30 09:16:17,500 2133688 WARNING cbx_2024-07-30_v16 odoo.modules.loading: Transient module states were reset
2024-07-30 09:16:17,501 2133688 ERROR cbx_2024-07-30_v16 odoo.modules.registry: Failed to load registry
Traceback (most recent call last):
  File "/opt/odoo/src/odoo/odoo/modules/registry.py", line 90, in new
    odoo.modules.load_modules(registry, force_demo, status, update_module)
  File "/opt/odoo/src/odoo/odoo/modules/loading.py", line 484, in load_modules
    processed_modules += load_marked_modules(cr, graph,
  File "/opt/odoo/src/odoo/odoo/modules/loading.py", line 372, in load_marked_modules
    loaded, processed = load_module_graph(
  File "/opt/odoo/src/odoo/odoo/modules/loading.py", line 183, in load_module_graph
    migrations.migrate_module(package, 'pre')
  File "/opt/odoo/local/OpenUpgrade/openupgrade_framework/odoo_patch/odoo/modules/migration.py", line 18, in migrate_module
    MigrationManager.migrate_module._original_method(self, pkg, stage)
  File "/opt/odoo/src/odoo/odoo/modules/migration.py", line 189, in migrate_module
    migrate(self.cr, installed_version)
  File "/srv/odoo/venv/lib/python3.10/site-packages/openupgradelib/openupgrade.py", line 2292, in wrapped_function
    func(
  File "/opt/odoo/local/addons-oca-community/addons/account_fiscal_year_closing/migrations/16.0.1.0.0/pre-migration.py", line 31, in migrate
    ("asset_receivable", env.ref("account.data_account_type_receivable").id),
  File "/opt/odoo/src/odoo/odoo/api.py", line 600, in ref
    record = self[res_model].browse(res_id)
  File "/opt/odoo/local/OpenUpgrade/openupgrade_framework/odoo_patch/odoo/api.py", line 39, in __getitem__
    return Environment.__getitem__._original_method(self, model_name)
  File "/opt/odoo/src/odoo/odoo/api.py", line 550, in __getitem__
    return self.registry[model_name](self, (), ())
  File "/opt/odoo/src/odoo/odoo/modules/registry.py", line 190, in __getitem__
    return self.models[model_name]
KeyError: 'account.account.type'
2024-07-30 09:16:17,501 2133688 CRITICAL cbx_2024-07-30_v16 odoo.service.server: Failed to initialize database `cbx_2024-07-30_v16`.
Traceback (most recent call last):
  File "/opt/odoo/src/odoo/odoo/service/server.py", line 1310, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "<decorator-gen-16>", line 2, in new
  File "/opt/odoo/src/odoo/odoo/tools/func.py", line 87, in locked
    return func(inst, *args, **kwargs)
  File "/opt/odoo/src/odoo/odoo/modules/registry.py", line 90, in new
    odoo.modules.load_modules(registry, force_demo, status, update_module)
  File "/opt/odoo/src/odoo/odoo/modules/loading.py", line 484, in load_modules
    processed_modules += load_marked_modules(cr, graph,
  File "/opt/odoo/src/odoo/odoo/modules/loading.py", line 372, in load_marked_modules
    loaded, processed = load_module_graph(
  File "/opt/odoo/src/odoo/odoo/modules/loading.py", line 183, in load_module_graph
    migrations.migrate_module(package, 'pre')
  File "/opt/odoo/local/OpenUpgrade/openupgrade_framework/odoo_patch/odoo/modules/migration.py", line 18, in migrate_module
    MigrationManager.migrate_module._original_method(self, pkg, stage)
  File "/opt/odoo/src/odoo/odoo/modules/migration.py", line 189, in migrate_module
    migrate(self.cr, installed_version)
  File "/srv/odoo/venv/lib/python3.10/site-packages/openupgradelib/openupgrade.py", line 2292, in wrapped_function
    func(
  File "/opt/odoo/local/addons-oca-community/addons/account_fiscal_year_closing/migrations/16.0.1.0.0/pre-migration.py", line 31, in migrate
    ("asset_receivable", env.ref("account.data_account_type_receivable").id),
  File "/opt/odoo/src/odoo/odoo/api.py", line 600, in ref
    record = self[res_model].browse(res_id)
  File "/opt/odoo/local/OpenUpgrade/openupgrade_framework/odoo_patch/odoo/api.py", line 39, in __getitem__
    return Environment.__getitem__._original_method(self, model_name)
  File "/opt/odoo/src/odoo/odoo/api.py", line 550, in __getitem__
    return self.registry[model_name](self, (), ())
  File "/opt/odoo/src/odoo/odoo/modules/registry.py", line 190, in __getitem__
    return self.models[model_name]
KeyError: 'account.account.type'
```

We believe that there is a timing issue with this pre-migration script, because at this stage, those records are already deleted and therefore the keyerror occurs

Borruso replied:
update pre-migration
can you check again?

OKT96 commented on account_fiscal_year_closing/migrations/16.0.1.0.0/pre-migration.py:
```diff
@@ -0,0 +1,61 @@
+            UPDATE {table} aa
+            SET account_type = atm.account_type
+            FROM account_type_map atm
+            WHERE atm.user_type_id = aa.user_type_id
```
```
ERROR: column aa.user_type_id does not exist
LINE 35:             WHERE atm.user_type_id = aa.user_type_id
                                              ^
HINT:  Perhaps you meant to reference the column "atm.user_type_id".
```

user_type_id does not exist in any of the two tables? (account_fiscalyear_closing_type_template, account_fiscalyear_closing_type)

Borruso replied:
yes field user_type_id become account_type

https://github.com/odoo/odoo/commit/26b2472f4977ccedbb0b5ed5f08be2c04313fd21 `user_type_id` -> `account_type` <del>`account.account.type`</del>

SirAionTech reviewed (commented):
I proposed https://github.com/Borruso/account-closing/pull/1 to support the flow where accounting data is created without loading a template chart of accounts

peterromao commented:
Hello guys,

Where can I translate this module for version 16? It's not on Weblate. Also as I see it it is not yet merged into the 16.0 branch either right?

andreampiovesana commented:
can you enable runboat?
is good for functional review and/or merge?

Borruso commented:
> Please include [55998e8](https://github.com/OCA/account-closing/commit/55998e8fc84b93764a66df09347467ae19add7e3) in this PR.

commit included

MaurizioPellegrinet reviewed (approved):
Functional test: OK

AinohaBH reviewed (approved):
Functional: LGTM

andreampiovesana reviewed (approved):
LGTM, MERGE?

salvorapi reviewed (approved):
LGTM

Borruso commented:
@pedrobaeza merge?

PS
After merge I will create PR for v17 and v18

pedrobaeza commented:
Well, I discourage the use of this module, and I don't think it's really needed in Italy, but merging due to reviews.

/ocabot merge nobump

lk-eska commented:
> Well, I discourage the use of this module, and I don't think it's really needed in Italy, but merging due to reviews.
> 
> /ocabot merge nobump

then how do we achieve the closing process? would you point us to a direction? thanks. 
@pedrobaeza

pedrobaeza commented:
Legislations usually don't talk about any "closing process". They just ask for financial books. The closing process is an old artificial method for being able to obtain that books with the limited resources that exist some time ago. A lot of people say the same that "systems have to make the closing", but we are working since a lot in Odoo without it, and other systems like SAP or Navision do the same.

lk-eska commented:
I believe people insist using this module because this is a legal requirement in their country (we are also in this boat).

pedrobaeza commented:
That's what I think that it's not really required in your country...
